src/assign.c

```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+            const SEXP val = sourceD[i];
+            newSourceD[i] = val==NA_STRING ? NA_INTEGER : -TRUELENGTH(val);
           }
         }
-        savetl_end();
-        setAttrib(target, R_LevelsSymbol, targetLevels=growVector(targetLevels, nTargetLevels + nAdd));
-        for (int k=0; k<nAdd; ++k) {
-          SET_STRING_ELT(targetLevels, nTargetLevels+k, temp[k]);
+        source = newSource;
+        for (int k=0; k<nTargetLevels; ++k) SET_TRUELENGTH(targetLevelsD[k], 0);  // don't need those anymore
+        if (nAdd) {
+          // cannot grow the levels yet as that would be R call which could fail to alloc and we have no hook to clear up
+          SEXP *temp = (SEXP *)malloc(nAdd * sizeof(SEXP *));
+          if (!temp) {
+            // # nocov start
+            for (int k=0; k<nSourceLevels; ++k) SET_TRUELENGTH(sourceLevelsD[k], 0);
+            savetl_end();
+            error("Unable to allocate working memory of %d bytes to combine factor levels", nAdd*sizeof(SEXP *));
+            // # nocov end
+          }
+          for (int k=0, thisAdd=0; thisAdd<nAdd; ++k) {   // thisAdd<nAdd to stop early when the added ones are all reached
+            SEXP s = sourceLevelsD[k];
+            int tl = TRUELENGTH(s);
+            if (tl) {  // tl negative here
+              if (tl != -nTargetLevels-thisAdd-1) error("Internal error: extra level check sum failed"); // # nocov
+              temp[thisAdd++] = s;
+              SET_TRUELENGTH(s,0);
+            }
+          }
+          savetl_end();
+          setAttrib(target, R_LevelsSymbol, targetLevels=growVector(targetLevels, nTargetLevels + nAdd));
+          for (int k=0; k<nAdd; ++k) {
+            SET_STRING_ELT(targetLevels, nTargetLevels+k, temp[k]);
+          }
+          free(temp);
+        } else {
+          // all source levels were already in target levels, but not with the same integers; we're done
+          savetl_end();
         }
-        free(temp);
-      } else {
-        // all source levels were already in target levels, but not with the same integers; we're done
-        savetl_end();
+        // now continue, but with the mapped integers in the (new) source
+      }
+    }
+  } else if (TYPEOF(target)!=TYPEOF(source)) {
+    // checks up front, otherwise we'd need checks twice in the two branches that cater for 'where' or not
+    bool pun = false;
+    switch(TYPEOF(target)) {
+    case LGLSXP:
+      if (isInteger(source)) {
+        const int *sD = INTEGER(source);
+        for (int i=0; i<slen; ++i) {
+          const int val = sD[i];
+          if (val!=0 && val!=1 && val!=NA_INTEGER) error("Cannot assign %d to a logical column.", val);
+        }
+        pun = true;
+      } else if (isReal(source)) {   // TODO: cater for integer64 here
+        const double *sD = REAL(source);
+        for (int i=0; i<slen; ++i) {
+          const double d = sD[i];
+          if (!ISNAN(d) && d!=0.0 && d!=1.0) error("Cannot assign %f to a logical column.", d);
```

Comment:
in the earlier branch you did the NA check last, I guess that will be more efficient in general, do that here as well?

Reply:
what I had in mind was that if d is nan, then !=0.0 returns nan not true.  But good point. Might be worth coming back to.   ISNAN in R is defined to be C isnan so it's a very fast compare.  I doubt we'd be able to get a measurable difference here -- it's all register ops.